gatsby/content/blog/2021/12/2021-12-03-sydent-typing-1.mdx

```diff
@@ -0,0 +1,187 @@
+---
+  date: '2021-12-03'
+  title: 'Type coverage for Sydent: motivation'
+  categories:
+    - Thoughts
+    - Tech
+  author: David Robertson
+---
+
+[Sydent](https://github.com/matrix-org/sydent) is the reference Matrix Identity server. It provides a [lookup service](https://matrix.org/docs/spec/identity_service/r0.3.0), so that you can find a Matrix user via their email address or phone number (if they've chosen to share it).
+
+We recently worked on [improving Sydent's type coverage](https://github.com/matrix-org/sydent/issues/414). Type coverage (or "typing") refers to annotations in Python source code that specify the type of an expression. These annotations are entirely optional, but if present, they allow tools to reason about the types that flow through a program. One popular tool is [`mypy`](https://mypy.readthedocs.io/en/stable/), which can often use the type information to spot bugs before you ship them! We worked to make Sydent pass `mypy --strict`. It now does so, with an overall precision of 94%! Here's what that process looks like:
+
+
+![Coverage as measured by mypy. Precision and the number of typed expressions increase over the latter half of 2021.](/blog/img/TWGXKTFpjYxlbnHRkZFAzljQ.png)
+
+Two lines show two different measures of how well-typed the project is. The grey region covers our two-week sprint towards improving coverage; the earliest data point is from just before previous efforts to improve typing earlier in the year.
+
+In a series of posts, I'd like to reflect on this sprint and share what we've learned. In particular, I aim to:
+
+- explain why we wanted to improve type coverage _now_;
+- work through examples to see how (if?) mypy could have spotted bugs;
+- describe the annotation process;
+- illustrate common patterns I learned along the way;
+- discuss the checks that `mypy` provides; and finally
+- reflect on the state of Python's typing ecosystem.
+
+In this first part, we'll concentrate on the first two topics; the [second](2021-12-10-sydent-typing-2.mdx) covers the middle two; and the [third](2021-12-17-sydent-typing-3.mdx) the last two.
```

Comment:
These will 404 at time of publication...

Reply:
I was hoping to be lazy. d6a48bcb does it properly.